src/ts/core/features/filter-search.ts

```diff
@@ -0,0 +1,130 @@
+}
+
+const checkSettingsAndSetupFilterSearchToggle = () => {
+    Settings.isActive('filter_search').then(active => {
+        toggleFilterSearch(active)
+    })
+}
+checkSettingsAndSetupFilterSearchToggle()
+
+browser.runtime.onMessage.addListener(async message => {
+    if (message === 'settings-updated') {
+        checkSettingsAndSetupFilterSearchToggle()
+    }
+})
+
+const toggleFilterSearch = (active: boolean) => {
+    if (active) {
+        document.addEventListener('click', filterSearch)
+    } else {
+        document.removeEventListener('click', filterSearch)
+    }
+}
+
+const searchInputId = 'searchInputBox'
+
+const getSearchInput = (): HTMLInputElement | null => {
+    return document.getElementById(searchInputId) as HTMLInputElement
+}
+
+async function filterSearch(event: MouseEvent) {
+    const eventTarget = event.target as HTMLElement
+    if (!eventTarget) return
+
+    const filterPageButtonClicked = eventTarget.className === 'bp3-button'
+    const filterActivationButtonClicked =
+        eventTarget.className === 'bp3-icon bp3-icon-filter' ||
+        eventTarget.className === 'bp3-button bp3-minimal bp3-small'
+
+    if (filterPageButtonClicked) {
+        refocusSearchInput()
+    } else if (filterActivationButtonClicked) {
+        await createSearchFilterElements()
+    }
+}
+
+function refocusSearchInput() {
+    const searchInput = getSearchInput()
+    if (searchInput) {
+        searchInput.focus()
+        searchInput.select()
+        onFilterInput()
+    }
+}
+
+async function createSearchFilterElements() {
+    const filterContainer = await waitForSelectorToExist(
+        'div.bp3-transition-container.bp3-popover-enter-done div.bp3-popover-content > div > div'
+    )
+
+    function createSeparator() {
+        const newDivLine = document.createElement('div')
+        newDivLine.className = 'rm-line'
+        filterContainer.prepend(newDivLine)
+    }
+
+    createSeparator()
+
+    function createSearchElementsContainer() {
+        const container = document.createElement('div')
+        container.id = 'filterBoxSearch'
+        container.style.cssText = 'display:flex;padding:4px'
+        filterContainer.prepend(container)
+        return container
+    }
+
+    const searchElementsContainer = createSearchElementsContainer()
+
+    function createSearchLabel() {
+        const searchLabel = document.createElement('strong')
+        searchLabel.innerText = 'Search'
+        searchLabel.style.cssText = 'margin-right:10px'
+        searchElementsContainer.appendChild(searchLabel)
+    }
+
+    createSearchLabel()
+
+    function createSearchInput() {
+        const searchInput = document.createElement('input')
+        searchInput.value = ''
+        searchInput.id = searchInputId
+        searchInput.name = searchInputId
+        searchInput.style.cssText = 'width:200px;display:flex'
+        searchElementsContainer.appendChild(searchInput)
+
+        searchInput.addEventListener('input', onFilterInput)
```

Comment:
I think naming the consequence of event handlers (`showButtonsMatchingQuery`) is usually more helpful than naming the cause, because readers probably already know "this happens when a search is inputted", but they don't know "what happens, and why does this handler exist" yet